### PR TITLE
Feature: TH to generate freer monadic actions

### LINKED ISSUE #16: TemplateHaskell for convenience funtions
Let us have a look at the example code.

~~~ .hs
data Console s where
    PutStrLn :: String -> Console ()
    GetLine :: Console String

putStrLn :: Member Console r => String -> Eff r ()
putStrLn = send . PutStrLn

getLine :: Member Console r => Eff r String
getLine = send GetLine
~~~

It seems like `putStrLn` and `getLine` is very boilerplaty and could be generated by TH.

### PR DESCRIPTION
**Pull request by isovector**
This PR adds `Control.Monad.Freer.TH`, which contains TH functions capable of generating type signatures and declarations for lifting GADT constructors into `Eff`. 

Non-GADT datatypes are not supported, although this isn't a particularly difficult thing to add support for later.

Closes #16

### CODE REVIEW

**Comment by trskop** [comment]
I've rebased the PR to current `master`, which is a good starting point for us to being able to accept it.

This will probably won't be able to compile on older GHC versions due to library incompatibilities. Hopefully at least GHC 8.0.* will be able to compile it.

Next steps:

* Resolve library incompatibilities.
* Write few tests for it.
* Enhance documentation with few more usage examples, and add at least one into `examples/`.

File: package.yaml [hunk 1]
```diff
@@ -60,6 +60,7 @@ ghc-options:
 
 dependencies:
   - base >=4.7 && <5
+  - template-haskell >=2.9 && <2.12
```
Comment (liskin) [inline package.yaml]: This should go into library dependencies only, I think.

Comment (trskop) [reply package.yaml]: Good point. Thanks for pointing it out.

**Comment by isovector** [comment]
What do you mean by resolving library incompatibilities? CPP to hide this from old versions of template-haskell? 

I'm by no means a TH expert, but what would tests for this stuff look like? Just "write a GADT, use TH to make actions, then then write tests using those actions", or do you have something more in mind?

**Comment by xkollar** [comment]
@isovector: see CI results...

**Comment by isovector** [comment]
@xkollar thanks! missed that somehow.

Looks like fundamentally `template-haskell` doesn't provide access to language extensions until 2.11.0.0, so there is no "real" way of solving this issue for the failing cases. What's the correct solution moving forwards? Bumping the TH lower bounds and hiding this module behind a cabal flag?

**Comment by trskop** [comment]
I think that checking if `FlexibleContexts` extension is enabled provides better error message than what GHC would print out normally. However, that doesn't mean we should disable this functionality entirely on older GHCs (i.e. older `template-haskell`).

**Comment by trskop** [comment]
Guys (in non-gender specific way), what do you think about the names `makeFreer`, and `makeFreer_`? Do we have better alternatives? IMHO, they aren't very descriptive in regards to their functionality.

**Comment by isovector** [comment]
The problem is unfortunately deeper than just the `FlexibleContexts` check -- also the `GadtC` constructor doesn't exist until 2.11.0.0, without which none of this works whatsoever :)

re: `makeFreer` -- the naming is intentionally analogous to https://hackage.haskell.org/package/free-4.12.4/docs/Control-Monad-Free-TH.html

**Comment by trskop** [comment]
Yes, that is a problem (missing GADT support), but we also need to generalise the TH code to support standard non-GADTs. There is a non-trivial class of effects that can be created without the need to pass type equalities around. I'd rather provide some TH code for older GHCs then none. It gets really confusing when Hackage says that you have a TH module, but the compiler says otherwise.

Regarding naming convention. IMHO, it's not very convincing that it's the same naming convention as `free` uses. Free applicatives and monads have nothing to do with Freer Effects, and especially effect types them selves. Convention used in e.g. [lens](https://hackage.haskell.org/package/lens-4.15.1/docs/Control-Lens-TH.html) makes much more sense to me. What would you say to something like `makeEffect`?

File: src/Control/Monad/Freer/TH.hs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+
+------------------------------------------------------------------------------
+-- | Like 'makeFreer', but does not provide type signatures.
+-- This can be used to attach Haddock comments to individual arguments
+-- for each generated function.
+--
+-- @
+-- data Lang x where
+--   Output :: String -> Lang x
```
Comment (Voltir) [inline src/Control/Monad/Freer/TH.hs]: Typo, should be: String -> Lang ()

**Comment by isovector** [comment]
FWIW, I'd prefer to get this merged, possibly with <2.11.0.0 CPP that `fail "this module requires template-haskell-2.11.0.0 (for now)"`s. This module as it stands has value, and after this week I'm not going to have much time to devote towards OSS until May. Thoughts?

File: src/Control/Monad/Freer/TH.hs (hunk 2 above)
Comment (isovector) [reply src/Control/Monad/Freer/TH.hs]: nice catch!

**Comment by xkollar** [comment]
Regarding naming of those `makeFreer` functions: I find naming convention of analogous functions in `free` arbitrary and thus do not see a reason to follow it...

Those functions do not create effects, they just create convenience functions... I would suggest `{make,mk}EffectCalls`, or something...

Naming things is hard... <https://martinfowler.com/bliki/TwoHardThings.html>

**Comment by trskop** [comment]
If you won't mind then we could take care of it for you. It may take a few days, since we are investing our free time into this.

BTW, thanks for doing this. It really helps to have such a great feedback.

**Comment by isovector** [comment]
Feel free to carry on! It's awesome to have such active maintainers of this package :)

**Comment by isovector** [comment]
Any word on this? I'm giving a talk on freer-effects next week at bayhac, and would love to be able to say "too much boilerplate for you? fear not! there's TH to help!" :)

**Comment by xkollar** [comment]
@isovector: Currently we are little tight on time but have not forgotten about it. Sorry for the inconvenience.

**Comment by mmaz** [comment]
any chance this could be merged in before the next release? thanks!!

**Comment by ProofOfKeags** [comment]
Is this happening?

**Comment by liskin** [comment]
Nothing is happening, we should never have forked this. Sorry. :-(

**Comment by ProofOfKeags** [comment]
Is it just that they aren’t being accepted upstream or is there some other
problem?

**Comment by liskin** [comment]
@CaptJakk I'm (one of) the upstream. I am the one who forked this 2 years ago, naïvely believing we'd have the resources to keep the project alive. We have all left the company since then. I'm no longer doing anything Haskell related in my day job, and I have way too much other projects in my spare time. My former colleagues do, as far as I know, work in Haskell-related positions, but apparently none of that involves freer-effects either.

This project is dead, let's admit it. I'm sorry.

**Comment by ProofOfKeags** [comment]
Got it. Is it just freer-effects that's dead or is freer dead as well?

**Comment by liskin** [comment]
@CaptJakk freer seems dead as well, that's why we forked it then.

**Comment by isovector** [comment]
@CaptJakk try https://github.com/lexi-lambda/freer-simple

**Pull request closed without merging**